torch/_inductor/freezing.py

```diff
@@ -224,3 +230,36 @@ def discard_traced_gm_params(mod):
             e_t.requires_grad_(True)
             e_t._is_param = True
         setattr(mod, attr_name, e_t)
+
+
+@dynamo_timed
+def convert_conv_weights_to_channels_last(gm):
+    """
+    Convert 4d convolution weight tensor to channels last format.
+
+    This method assumes the graph is already freezed.
+    """
+    convs = [n for n in gm.graph.nodes if n.target == aten.convolution.default]
+    for conv in convs:
+        weight_node = conv.args[1]
+        # is a constant tensor
+        if weight_node.op == "get_attr":
+            param_tensor = getattr(gm, weight_node.target)
+            if len(param_tensor.shape) != 4:
+                # not a 4d tensor, skip
+                continue
+            with no_dispatch():
+                cl_param_tensor = param_tensor.to(memory_format=torch.channels_last)
+                if isinstance(param_tensor, nn.Parameter):
+                    cl_param_tensor = nn.Parameter(cl_param_tensor)
+            if cl_param_tensor is not param_tensor:
+                setattr(gm, weight_node.target, cl_param_tensor)
+
+                # Even though inductor does not use meta['val'] or meta['tensor_meta']
+                # for get_attr node, we still update them to be consistent.
+                weight_node.meta["val"] = weight_node.meta["val"].to(
```

Comment:
We may not use ```meta['val']``` or ```meta['tensor_meta']``` for the get_attr node, we change the weight format here, the convolution node infos may also need to be changed or we don't need to care about it?

Reply:
I think if we want the meta data in every node of the graph to be consistent with the layout conversion, we need do a full pass thru the graph and update potentially every reached node's layout. But inductor may make different decision to the saved layout (in meta['val']) anyways, so making meta['val'] for intermediate nodes up to date sounds like an over-kill.  Leave them reflecting the original eager layout pretending no layout conversion looks not too bad.

Reply:
It might be easier to just inset the weight conversions, and then retrace to get the updated .meta['val']. 

Changing weights to channels last could affect more than just the weight_node meta, other nodes as well.

Reply:
Hmm, I still think we should not do it ..

Layout optimization is something not visible to user hopefully. Most of the time .meta['val'] represents the eager stride *without* layout optimization. If we retrace with CL inputs,  then what we end up have in .meta['val'] are those strides *with* layout optimizations.  For intermediate tensors, this is fine since they are not user visible, but it's bad for output tensors. We have been take care to maintain the same stride as eager for output tensors, we probably should not break it.